### PR TITLE
Set state with heartbeat implementation

### PR DESCRIPTION
**Pull request by mateusz-dahlke**

### CODE REVIEW

File: core/pubnub_coreapi_ex.c [hunk 1]
```diff
@@ -241,3 +242,60 @@ enum pubnub_res pubnub_history_ex(pubnub_t*                     pb,
...
+        pubnub_mutex_unlock(pb->monitor);
+        return PNR_IN_PROGRESS;
+    }
+
+    rslt = pbauto_heartbeat_prepare_channels_and_ch_groups(pb, &channel, &opts.channel_group);
+    if (rslt != PNR_OK) {
+        return rslt;
+    }
+
+    pb->core.state = state;
```
Comment (budgetpreneur) [inline core/pubnub_coreapi_ex.c:277]: I believe this is incorrect way to set state or inconsistent way from the functionality of set_state
Please check https://github.com/pubnub/c-core/blob/master/core/pubnub_coreapi.c#L319-L408

Comment (Xavrax) [reply core/pubnub_coreapi_ex.c:277]: ahhh I see. Okay thanks! Gonna change that.

File: core/pubnub_coreapi_ex.c [hunk 2]
```diff
@@ -273,15 +273,22 @@ static enum pubnub_res heartbeat_with_state(pubnub_t *pb,
     if (rslt != PNR_OK) {
         return rslt;
     }
-
-    pb->core.state = state;
+    
+    size_t state_lenght = strlen(state);
```
Comment (budgetpreneur) [inline core/pubnub_coreapi_ex.c:277]: variable name typo? state_lenght...may be it should be state_length

File: core/pubnub_coreapi_ex.c [hunk 3]
```diff
@@ -241,3 +242,67 @@ enum pubnub_res pubnub_history_ex(pubnub_t*                     pb,
...
+        return rslt;
+    }
+    
+    size_t state_lenght = strlen(state);
+    if (pb->core.state == NULL) {
+        pb->core.state = (char*)malloc(state_lenght + 1);
+    } else if (strlen(pb->core.state) < state_lenght) {
+        pb->core.state = (char*)realloc((char*)pb->core.state, state_lenght + 1);
+    }
+    strcpy((char*)pb->core.state, state);
```
Comment (budgetpreneur) [inline core/pubnub_coreapi_ex.c:283]: heartbeat=true, is not giving expected state in the URL.
with existing set_state, heartbeat URL is `/v2/presence/sub-key/demo/channel/hello_world/heartbeat?pnsdk=Windows-PubNub-C-core%2F4.0.6&uuid=6e581826-3787-4d70-a465-d749226345f5&state=%7B%22hello_world%22%3A%7B%22x%22%3A5%7D%7D&auth=danaske`
Note the format of state {<channel>:<state provided by user>}

with heartbeat=true, the URL I see is `/v2/presence/sub-key/demo/channel/hello_world/heartbeat?pnsdk=Windows-PubNub-C-core%2F4.0.6&uuid=91fa116f-871b-459a-9009-b3997eb6b041&state=%7B%22x%22%3A5%7D&auth=danaske`
which is missing channel name in the state.

File: core/pubnub_coreapi_ex.c (hunk 2 above)
Comment (Xavrax) [reply core/pubnub_coreapi_ex.c]: fixed

File: core/pubnub_coreapi_ex.c (hunk 3 above)
Comment (Xavrax) [reply core/pubnub_coreapi_ex.c]: Changed as we discuss at DM

File: core/Makefile [hunk 4]
```diff
@@ -46,7 +47,7 @@ CFLAGS +=-g -D PUBNUB_ADVANCED_KEEP_ALIVE=1 -D PUBNUB_LOG_LEVEL=PUBNUB_LOG_LEVEL
 
 LDFLAGS=-L../cgreen/build/src
 
-CGREEN_RUNNER=../cgreen/build/tools/cgreen-runner
+CGREEN_RUNNER=../../../cgreen/build/tools/cgreen-runner
```
Comment (budgetpreneur) [inline core/Makefile:50]: is this ok? current heartbeat related changes should not change CGREEN stuff.

File: core/pubnub_core_unit_test.c [hunk 5]
```diff
@@ -8,6 +8,8 @@
 #include "pubnub_server_limits.h"
 #include "pubnub_pubsubapi.h"
 #include "pubnub_coreapi.h"
+// TODO: move coreapi_ex to new module
```
Comment (budgetpreneur) [inline core/pubnub_core_unit_test.c:11]: TODO still valid?

Comment (budgetpreneur) [reply core/pubnub_core_unit_test.c:11]: oh...this is test file..Never mind..Please ignore.

File: core/pubnub_coreapi.c [hunk 6]
```diff
@@ -7,6 +7,8 @@
 #include "pubnub_assert.h"
 #include "pubnub_timers.h"
 #include "pubnub_memory_block.h"
+// TODO: why not all pbcc?
```
Comment (budgetpreneur) [inline core/pubnub_coreapi.c:10]: is TODO still applicable?

**Review by budgetpreneur — APPROVED** [review]
LGTM other than the minor comments I gave.

File: core/pubnub_coreapi.c (hunk 6 above)
Comment (mateusz-dahlke) [reply core/pubnub_coreapi.c:10]: Yeah. It was my question but I think that's C sourcing stuff and I think that I don't want to create a discussion about it. I will remove that todo.

File: core/Makefile (hunk 4 above)
Comment (mateusz-dahlke) [reply core/Makefile:50]: good caught. I do this for launching my tests. Thanks!

**Pull request merged**